### PR TITLE
[RISCV] Make X5 allocatable for JALR on CPUs without RAS

### PR DESCRIPTION
**Pull request by wangpc-pp**
Some microarchitectures may not support RAS, then we don't need to
reserve X5 register for JALR.

If RAS is supported, we will select the register allocation order
without X5 (because alternative orders should be subsets of the
default order).

### CODE REVIEW

File: llvm/lib/Target/RISCV/RISCVFeatures.td [hunk 1]
```diff
@@ -970,6 +970,9 @@ def FeatureFastUnalignedAccess
 def FeaturePostRAScheduler : SubtargetFeature<"use-postra-scheduler",
     "UsePostRAScheduler", "true", "Schedule again after register allocation">;
 
+def FeatureNoRAS : SubtargetFeature<"no-ras", "HasRAS", "false",
+                                    "Hasn't RAS (Return Address Stack)">;
+
```
Comment (dtcxzyw) [inline llvm/lib/Target/RISCV/RISCVFeatures.td:975]: https://github.com/llvm/llvm-project/blob/8f7fdd94ef19af7b4905b316c253a78219a6038f/llvm/lib/Target/ARM/ARM.td#L418-L422

Not sure which one is better.

**Comment by topperc** [comment]
Do you have any performance or code size data to show that this change is a benefit to CPUs with RAS?

**Comment by wangpc-pp** [comment]
> Do you have any performance or code size data to show that this change is a benefit to CPUs with RAS?

I don't have performance data as I don't have such hardware implementation. And I don't see code size change on llvm-test-suite.
The thought of this PR came from a random discussion with my colleague, and ARM supports such feature. So I think maybe some low-end products need it.

File: llvm/lib/Target/RISCV/RISCVFeatures.td (hunk 1 above)
Comment (wangpc-pp) [reply llvm/lib/Target/RISCV/RISCVFeatures.td:975]: Yeah, RAS can be ambiguity as RAS can be `Reliability, Availability, Serviceability`.
I will align it to ARM later.

**Comment by topperc** [comment]
> > Do you have any performance or code size data to show that this change is a benefit to CPUs with RAS?
> 
> I don't have performance data as I don't have such hardware implementation. And I don't see code size change on llvm-test-suite. The thought of this PR came from a random discussion with my colleague, and ARM supports such feature. So I think maybe some low-end products need it.

It doesn't look like ARM uses it for register allocation though. This is the only code I can find on ARM that uses it

```
    if (!isDirect && !Subtarget->hasV5TOps())                                    
      CallOpc = ARMISD::CALL_NOLINK;                                             
    else if (doesNotRet && isDirect && Subtarget->hasRetAddrStack() &&           
             // Emit regular call when code size is the priority                 
             !Subtarget->hasMinSize())                                           
      // "mov lr, pc; b _foo" to avoid confusing the RSP                         
      CallOpc = ARMISD::CALL_NOLINK;                                             
    else                                                                         
      CallOpc = isLocalARMFunc ? ARMISD::CALL_PRED : ARMISD::CALL;
```

**Comment by wangpc-pp** [comment]
> > > Do you have any performance or code size data to show that this change is a benefit to CPUs with RAS?
> > 
> > 
> > I don't have performance data as I don't have such hardware implementation. And I don't see code size change on llvm-test-suite. The thought of this PR came from a random discussion with my colleague, and ARM supports such feature. So I think maybe some low-end products need it.
> 
> It doesn't look like ARM uses it for register allocation though. This is the only code I can find on ARM that uses it
> 
> ```
>     if (!isDirect && !Subtarget->hasV5TOps())                                    
>       CallOpc = ARMISD::CALL_NOLINK;                                             
>     else if (doesNotRet && isDirect && Subtarget->hasRetAddrStack() &&           
>              // Emit regular call when code size is the priority                 
>              !Subtarget->hasMinSize())                                           
>       // "mov lr, pc; b _foo" to avoid confusing the RSP                         
>       CallOpc = ARMISD::CALL_NOLINK;                                             
>     else                                                                         
>       CallOpc = isLocalARMFunc ? ARMISD::CALL_PRED : ARMISD::CALL;
> ```

This is some kind of reversion of your patch: https://reviews.llvm.org/D105875, it seems that we used to use GPR for JALR. The benefit is that such implementaion can have one more available register.
I don't have more data to support such configuration, so I'll let you guys decide whether we should support it. :-)

File: llvm/test/CodeGen/RISCV/calls.ll [hunk 2]
```diff
@@ -1,115 +1,74 @@
...
-; RV32I-PIC-NEXT:    addi sp, sp, 16
-; RV32I-PIC-NEXT:    ret
+; CHECK-LABEL: test_call_external:
+; CHECK:       # %bb.0:
+; CHECK-NEXT:    addi sp, sp, -16
+; CHECK-NEXT:    sw ra, 12(sp) # 4-byte Folded Spill
+; CHECK-NEXT:    call external_function
+; CHECK-NEXT:    lw ra, 12(sp) # 4-byte Folded Reload
+; CHECK-NEXT:    addi sp, sp, 16
+; CHECK-NEXT:    ret
```
Comment (kito-cheng) [inline llvm/test/CodeGen/RISCV/calls.ll:19]: This change seems like a NFC, also that is not really related to this change, so maybe a separated NFC patch?

**Comment by wangpc-pp** [comment]
Ping for comments. :-)

**Comment by topperc** [comment]
> Ping for comments. :-)

If this didn't improve code size on llvm-test-suite, I think it just adds an extra configuration option for no benefit.

The ARM code specifically seems to use the feature to generate a longer code sequence for noreturn calls on CPUs that support RAS to avoid corrupting the predictor. So that seems different than this patch.

**Comment by wangpc-pp** [comment]
> > Ping for comments. :-)
> 
> If this didn't improve code size on llvm-test-suite, I think it just adds an extra configuration option for no benefit.
> 
> The ARM code specifically seems to use the feature to generate a longer code sequence for noreturn calls on CPUs that support RAS to avoid corrupting the predictor. So that seems different than this patch.

Yeah, make sense to me. We may investigate this again if it will be profitable in the future.

**Comment by dtcxzyw** [comment]
> > Ping for comments. :-)
> 
> If this didn't improve code size on llvm-test-suite, I think it just adds an extra configuration option for no benefit.
> 

Result: https://github.com/dtcxzyw/llvm-ci/pull/1006#issuecomment-1926433112

**Comment by wangpc-pp** [comment]
> > > Ping for comments. :-)
> > 
> > 
> > If this didn't improve code size on llvm-test-suite, I think it just adds an extra configuration option for no benefit.
> 
> Result: [dtcxzyw/llvm-ci#1006 (comment)](https://github.com/dtcxzyw/llvm-ci/pull/1006#issuecomment-1926433112)

Thanks! The result is the same as my previous.
